src/game/drawing.rs

```diff
@@ -153,69 +148,75 @@ fn goal_unlock_animation_system(
 	} else {
 		palette.goal_closed
 	};
-	for mut sprite in &mut query {
-		sprite.color = color;
+	for children in &flags_q {
+		for id in children {
+			if let Ok(mut sprite) = sprites_q.get_mut(*id) {
+				sprite.color = color;
+			}
+		}
 	}
 }
 
 fn button_trigger_animation_system(
-	mut buttons_q: Query<(&mut Sprite, Option<&LogicalColor>), With<BoxSlot>>,
-	mut boxes_q: Query<
-		(&mut Sprite, Option<&LogicalColor>),
-		(
-			With<Box>,
-			Without<BoxSlot>, /* To guarantee memory aliasing */
-		),
-	>,
+	mut sprites_q: Query<&mut Sprite>,
+	buttons_q: Query<(&Children, Option<&LogicalColor>), With<BoxSlot>>,
+	boxes_q: Query<(&Children, Option<&LogicalColor>), With<Box>>,
 	nodes_q: Query<(&PlacedGlyph, &PlacedObject)>,
 	palette: Res<ThingPalette>,
 ) {
+	let mut recolor_sprites = |sprite_ids: &Children, color| {
+		for id in sprite_ids {
+			if let Ok(mut sprite) = sprites_q.get_mut(*id) {
+				sprite.color = color;
+			}
+		}
+	};
+
 	for (glyph_id, object_id) in &nodes_q {
-		let button = glyph_id.0.and_then(|id| buttons_q.get_mut(id).ok());
-		let object = object_id.0.and_then(|id| boxes_q.get_mut(id).ok());
+		let button = glyph_id.0.and_then(|id| buttons_q.get(id).ok());
+		let object = object_id.0.and_then(|id| boxes_q.get(id).ok());
 		// Use trigger color if both things are at the same place
 		// and are of the same logical color, otherwise use base color
 		match (button, object) {
-			(Some(mut button), Some(mut object)) => {
-				let colors_compatible = object
-					.1
+			(Some((button, button_color)), Some((object, object_color))) => {
+				let colors_compatible = object_color
 					.and_then(|object_color| {
-						button.1.map(|glyph_color| *object_color == *glyph_color)
+						button_color.map(|glyph_color| *object_color == *glyph_color)
 					})
 					// If either thing is colorless, they are considered compatible
 					.unwrap_or(true);
```

Comment:
These evaluations should maybe be stored in a purely logical component.

(Edit: The colours should probably be just animated like everything else visual. That's for another PR then.)